packages/core/src/agent/didcomm/versions/v2/sicpa-didcomm/DIDCommV2EnvelopeService.ts

```diff
@@ -0,0 +1,95 @@
+import type { AgentContext } from '../../../../context'
+import type { DecryptedMessageContext, EncryptedMessage, SignedMessage } from '../../../types'
+import type { DIDCommV2Message } from '../DIDCommV2Message'
+import type { DIDCommV2EnvelopeService, PackMessageParams } from '../index'
+import type { default as didcomm, IMessage } from 'didcomm'
+
+import { Key } from '../../../../../crypto'
+import { AriesFrameworkError } from '../../../../../error/AriesFrameworkError'
+import { injectable } from '../../../../../plugins'
+import { JsonEncoder } from '../../../../../utils'
+import { AgentConfig } from '../../../../AgentConfig'
+import { DIDCommMessageVersion, MessageType } from '../../../types'
+
+import { DIDResolverService } from './DIDResolverService'
+import { SecretResolverService } from './SecretResolverService'
+
+@injectable()
+export class DIDCommV2SicpaEnvelopeService implements DIDCommV2EnvelopeService {
+  private didResolverService: DIDResolverService
+  private secretResolverService: SecretResolverService
+  private didcomm: typeof didcomm
+
+  public constructor(
+    agentConfig: AgentConfig,
+    didResolverService: DIDResolverService,
+    secretResolverService: SecretResolverService
+  ) {
+    this.didcomm = agentConfig.agentDependencies.didcomm
+    this.didResolverService = didResolverService
+    this.secretResolverService = secretResolverService
+  }
+
+  public async packMessage(
+    agentContext: AgentContext,
+    payload: DIDCommV2Message,
+    params: PackMessageParams
+  ): Promise<EncryptedMessage> {
+    const message = new this.didcomm.Message(payload.toJSON() as IMessage)
+
+    if (params.type === MessageType.Signed && params.signByDID) {
+      const [encryptedMsg] = await message.pack_signed(
+        params.signByDID,
+        this.didResolverService.bindAgentContext(agentContext),
+        this.secretResolverService
+      )
+      return JsonEncoder.fromString(encryptedMsg)
+    }
+    if ((params.type === MessageType.Encrypted || !params.type) && params.toDID) {
+      const [encryptedMsg] = await message.pack_encrypted(
+        params.toDID,
+        params.fromDID || null,
+        params.signByDID || null,
+        this.didResolverService.bindAgentContext(agentContext),
```

Comment:
Once the did resolver is registered as ContextScoped (registerContextScoped in module) we can inject it herE:

```
const sicpaDidResolver = agentContext.resolve(DIDResolverService)
```